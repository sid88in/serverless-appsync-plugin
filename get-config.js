const fs = require('fs');
const path = require('path');
const { mapObjIndexed, pipe, values, merge } = require('ramda');

const objectToArrayWithNameProp = pipe(
  mapObjIndexed((item, key) => merge({ name: key }, item)),
  values
);

module.exports = (config, provider, servicePath) => {
  // TODO verify authenticationType
  if (!config.authenticationType) {
    throw new Error('appSync property `authenticationType` is required.');
  }
  if (!config.serviceRole) {
    throw new Error('appSync property `serviceRole` is required.');
  }
  if (
    config.authenticationType === 'AMAZON_COGNITO_USER_POOLS' &&
    !config.userPoolConfig
  ) {
    throw new Error(
      'appSync property `userPoolConfig` is required when authenticationType `AMAZON_COGNITO_USER_POOLS` is chosen.'
    );
  }

  const mappingTemplatesLocation = config.mappingTemplatesLocation || 'mapping-templates'
  const mappingTemplates = config.mappingTemplates || []

  const schemaPath = path.join(servicePath, config.schema || 'schema.graphql');
  const schemaContent = fs.readFileSync(schemaPath, {
    encoding: 'utf8'
  });

  const dataSources = objectToArrayWithNameProp(config.dataSources);

  return {
    name: config.name || 'api',
    apiId: config.apiId,
    region: provider.region,
    authenticationType: config.authenticationType,
    schema: schemaContent,
    userPoolConfig: config.userPoolConfig,
    serviceRoleArn: config.serviceRole,
    // TODO verify dataSources structure
    dataSources,
    mappingTemplatesLocation,
    mappingTemplates
  };
};
