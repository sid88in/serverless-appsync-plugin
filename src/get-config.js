const fs = require('fs');
const path = require('path');
const { mergeTypeDefs } = require('@graphql-tools/merge');
const { mapObjIndexed, pipe, values, merge } = require('ramda');
const globby = require('globby');

const objectToArrayWithNameProp = pipe(
  mapObjIndexed((item, key) => merge({ name: key }, item)),
  values,
);

const mergeTypes = (types, options) => {
  const schemaDefinition =
    options && typeof options.schemaDefinition === 'boolean'
      ? options.schemaDefinition
      : true;
  return mergeTypeDefs(types, {
    useSchemaDefinition: schemaDefinition,
    forceSchemaDefinition: schemaDefinition,
    throwOnConflict: true,
    commentDescriptions: true,
    reverseDirectives: true,
    ...options,
  });
};

const getConfig = (config, provider, servicePath) => {
  if (
    !(
      config.authenticationType === 'API_KEY' ||
      config.authenticationType === 'AWS_IAM' ||
      config.authenticationType === 'AMAZON_COGNITO_USER_POOLS' ||
      config.authenticationType === 'OPENID_CONNECT' ||
      config.authenticationType === 'AWS_LAMBDA' ||
      config.apiId
    )
  ) {
    throw new Error(
      'appSync property `authenticationType` is missing or invalid.',
    );
  }
  if (
    config.authenticationType === 'AMAZON_COGNITO_USER_POOLS' &&
    !config.userPoolConfig
  ) {
    throw new Error(
      'appSync property `userPoolConfig` is required when authenticationType `AMAZON_COGNITO_USER_POOLS` is chosen.',
    );
  }
  if (
    config.authenticationType === 'AWS_LAMBDA' &&
    !config.lambdaAuthorizerConfig
  ) {
    throw new Error(
      'appSync property `lambdaAuthorizerConfig` is required when authenticationType `AWS_LAMBDA` is chosen.',
    );
  }
  if (
    config.authenticationType === 'OPENID_CONNECT' &&
    !config.openIdConnectConfig
  ) {
    throw new Error(
      'appSync property `openIdConnectConfig` is required when authenticationType `OPENID_CONNECT` is chosenXXX.',
    );
  }

  if (config.logConfig && !config.logConfig.level) {
    throw new Error(
      'logConfig property `level` must be NONE, ERROR, or ALL when logConfig exists.',
    );
  }
  if (config.substitutions && typeof config.substitutions !== 'object') {
    throw new Error('substitutions property must be an object');
  }
  if (config.xrayEnabled && typeof config.xrayEnabled !== 'boolean') {
    throw new Error('xrayEnabled must be a boolean');
  }

  const mappingTemplatesLocation =
    config.mappingTemplatesLocation || 'mapping-templates';
  const functionConfigurationsLocation =
    config.functionConfigurationsLocation || mappingTemplatesLocation;
  const functionConfigurations = config.functionConfigurations || [];
  const mappingTemplates = config.mappingTemplates || [];

  const toAbsolutePosixPath = (filePath) =>
    (path.isAbsolute(filePath)
      ? filePath
      : path.join(servicePath, filePath)
    ).replace(/\\/g, '/');
  const readSchemaFile = (filePath) =>
    fs.readFileSync(filePath, { encoding: 'utf8' });

  const schema = Array.isArray(config.schema)
    ? config.schema
    : [config.schema || 'schema.graphql'];
  const schemaFiles = [].concat(
    ...schema.map((s) => globby.sync(toAbsolutePosixPath(s))),
  );
  const schemaContent = mergeTypes(schemaFiles.map(readSchemaFile));

  let dataSources = [];
  if (Array.isArray(config.dataSources)) {
    dataSources = config.dataSources.reduce((acc, value) => {
      // Do not call `objectToArrayWithNameProp` on datasources objects``
      if (value.name !== undefined && typeof value.name === 'string') {
        return acc.concat(value);
      }
      return acc.concat(objectToArrayWithNameProp(value));
    }, []);
  } else {
    dataSources = objectToArrayWithNameProp(config.dataSources);
  }

  return {
    ...config,
    name: config.name || 'api',
    region: provider.region,
    authenticationType: config.authenticationType,
    additionalAuthenticationProviders:
      config.additionalAuthenticationProviders || [],
    schema: schemaContent,
    // TODO verify dataSources structure
    dataSources,
    defaultMappingTemplates: config.defaultMappingTemplates || {},
    mappingTemplatesLocation,
    mappingTemplates,
    functionConfigurationsLocation,
    functionConfigurations,
    substitutions: config.substitutions || {},
    xrayEnabled: config.xrayEnabled || false,
  };
};

module.exports = (config, provider, servicePath) => {
  if (!config) {
    return [];
  } else if (config.constructor === Array) {
    return config.map((apiConfig) =>
      getConfig(apiConfig, provider, servicePath),
    );
  }
  const singleConfig = getConfig(config, provider, servicePath);
  singleConfig.isSingleConfig = true;
  return [singleConfig];
};
