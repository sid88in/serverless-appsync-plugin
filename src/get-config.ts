import fs from 'fs';
import path from 'path';
import { mergeTypeDefs } from '@graphql-tools/merge';
import { has, mapObjIndexed, pipe, values } from 'ramda';
import globby from 'globby';
import {
  ApiKeyConfig,
  AppSyncConfig,
  Auth,
  DataSource,
  FunctionConfig,
  Resolver,
  WafRule,
} from './types/plugin';
import { AWS } from '@serverless/typescript';
import { convertAppSyncSchemas } from 'appsync-schema-converter';
import { IntrinsictFunction } from 'types/cloudFormation';

const objectToArrayWithNameProp = pipe(
  mapObjIndexed(
    (item: DataSource | Omit<DataSource, 'name'>, key) =>
      ({
        name: key,
        ...item,
      } as DataSource),
  ),
  values,
);

const readSchemaFile = (filePath: string) =>
  fs.readFileSync(filePath, { encoding: 'utf8' });

const mergeTypes = (types) => {
  return mergeTypeDefs(types, {
    useSchemaDefinition: true,
    forceSchemaDefinition: true,
    throwOnConflict: true,
    commentDescriptions: true,
    reverseDirectives: true,
  });
};

const buildAppSyncSchema = (schemaFiles: string[]) => {
  // Merge files
  const mergedSchema = mergeTypes(schemaFiles.map(readSchemaFile))
    .replace(/ *#+(.*)/g, '"""\n$1\n"""')
    .replace(/"""\n"""\n/, '');

  return convertAppSyncSchemas(mergedSchema);
};

export type AppSyncConfigInput = {
  apiId?: string;
  isSingleConfig?: boolean;
  name?: string;
  region: string;
  schema: string | string[];
  apiKeys?: ApiKeyConfig[];
  caching?: {
    behavior: 'FULL_REQUEST_CACHING' | 'PER_RESOLVER_CACHING';
    type?: string;
    ttl?: number;
    atRestEncryption?: boolean;
    transitEncryption?: boolean;
  };
  additionalAuthenticationProviders?: Auth[];
  logConfig?: {
    loggingRoleArn?: string | IntrinsictFunction;
    level?: 'ERROR' | 'NONE' | 'ALL';
    excludeVerboseContent?: boolean;
  };
  defaultMappingTemplates?: {
    request?: string | false;
    response?: string | false;
  };
  mappingTemplatesLocation?: string;
  functionConfigurationsLocation?: string;
  mappingTemplates?: Resolver[];
  functionConfigurations?: FunctionConfig[];
  dataSources:
    | (DataSource | Record<string, DataSource>)[]
    | Record<string, DataSource>;
  substitutions?: Record<string, string | IntrinsictFunction>;
  xrayEnabled?: boolean;
  wafConfig?: {
    enabled: boolean;
    name: string;
    defaultAction: 'Allow' | 'Block';
    description?: string;
    rules: WafRule[];
  };
  tags?: Record<string, string>;
} & Auth;

const getAppSyncConfig = async (
  config: AppSyncConfigInput,
  provider: AWS['provider'],
  servicePath: string,
): Promise<AppSyncConfig> => {
  if (
    !(
      config.apiId ||
      config.authenticationType === 'API_KEY' ||
      config.authenticationType === 'AWS_IAM' ||
      config.authenticationType === 'AMAZON_COGNITO_USER_POOLS' ||
      config.authenticationType === 'OPENID_CONNECT' ||
      config.authenticationType === 'AWS_LAMBDA'
    )
  ) {
    throw new Error(
      'appSync property `authenticationType` is missing or invalid.',
    );
  }

  if (
    config.authenticationType === 'AMAZON_COGNITO_USER_POOLS' &&
    !has('userPoolConfig', config)
  ) {
    throw new Error(
      'appSync property `userPoolConfig` is required when authenticationType `AMAZON_COGNITO_USER_POOLS` is chosen.',
    );
  }
  if (
    config.authenticationType === 'AWS_LAMBDA' &&
    !has('lambdaAuthorizerConfig', config)
  ) {
    throw new Error(
      'appSync property `lambdaAuthorizerConfig` is required when authenticationType `AWS_LAMBDA` is chosen.',
    );
  }
  if (
    config.authenticationType === 'OPENID_CONNECT' &&
    !has('openIdConnectConfig', config)
  ) {
    throw new Error(
      'appSync property `openIdConnectConfig` is required when authenticationType `OPENID_CONNECT` is chosenXXX.',
    );
  }

  if (config.logConfig && !config.logConfig.level) {
    throw new Error(
      'logConfig property `level` must be NONE, ERROR, or ALL when logConfig exists.',
    );
  }
  if (config.substitutions && typeof config.substitutions !== 'object') {
    throw new Error('substitutions property must be an object');
  }
  if (config.xrayEnabled && typeof config.xrayEnabled !== 'boolean') {
    throw new Error('xrayEnabled must be a boolean');
  }

  const mappingTemplatesLocation =
    config.mappingTemplatesLocation || 'mapping-templates';
  const functionConfigurationsLocation =
    config.functionConfigurationsLocation || mappingTemplatesLocation;
  const functionConfigurations: FunctionConfig[] =
    config.functionConfigurations || [];
  const mappingTemplates: Resolver[] = config.mappingTemplates || [];

  const toAbsolutePosixPath = (filePath: string) =>
    (path.isAbsolute(filePath)
      ? filePath
      : path.join(servicePath, filePath)
    ).replace(/\\/g, '/');

  const schema = Array.isArray(config.schema)
    ? config.schema
    : [config.schema || 'schema.graphql'];
  const schemaFiles = ([] as string[]).concat(
    ...schema.map((s) => globby.sync(toAbsolutePosixPath(s))),
  );

  const schemaContent = buildAppSyncSchema(schemaFiles);

  let dataSources: DataSource[] = [];
  if (Array.isArray(config.dataSources)) {
    dataSources = config.dataSources.reduce((acc, value) => {
      // Do not call `objectToArrayWithNameProp` on datasources objects``
      if (has('name')(value)) {
        return acc.concat(value);
      } else {
        return acc.concat(objectToArrayWithNameProp(value));
      }
    }, [] as DataSource[]);
  } else if (config.dataSources) {
    dataSources = objectToArrayWithNameProp(config.dataSources);
  }

  return {
    ...config,
    name: config.name || 'api',
    region: provider.region || 'us-east-1',
    additionalAuthenticationProviders:
      config.additionalAuthenticationProviders || [],
    schema: schemaContent,
    // TODO verify dataSources structure
    dataSources,
    defaultMappingTemplates: config.defaultMappingTemplates || {},
    mappingTemplatesLocation,
    mappingTemplates,
    functionConfigurationsLocation,
    functionConfigurations,
    substitutions: config.substitutions || {},
    xrayEnabled: config.xrayEnabled || false,
  };
};

export const getConfig = async (
  config: AppSyncConfigInput | AppSyncConfigInput[],
  provider: AWS['provider'],
  servicePath: string,
) => {
  if (!config) {
    return [];
  } else if (Array.isArray(config)) {
    const conf: AppSyncConfig[] = [];
    for (const key in config) {
      conf.push(await getAppSyncConfig(config[key], provider, servicePath));
    }
    return conf;
  }

  const singleConfig = await getAppSyncConfig(config, provider, servicePath);
  singleConfig.isSingleConfig = true;

  return [singleConfig];
};
