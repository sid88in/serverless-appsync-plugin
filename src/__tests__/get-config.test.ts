import { getAppSyncConfig, ResolverConfigInput } from '../get-config';
import { basicConfig } from './basicConfig';

test('returns basic config', async () => {
  expect(getAppSyncConfig(basicConfig)).toMatchSnapshot();
});

describe('Schema', () => {
  it('should return the default schema', () => {
    expect(
      getAppSyncConfig({ ...basicConfig, schema: undefined }).schema,
    ).toMatchSnapshot();
  });

  it('should return a single schema as an array', () => {
    expect(
      getAppSyncConfig({ ...basicConfig, schema: 'mySchema.graphql' }).schema,
    ).toMatchSnapshot();
  });

  it('should return a schema array unchanged', () => {
    expect(
      getAppSyncConfig({
        ...basicConfig,
        schema: ['users.graphql', 'posts.graphql'],
      }).schema,
    ).toMatchSnapshot();
  });
});

describe('DataSources', () => {
  it('should resolve dataSource names', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      dataSources: {
        dataSourceWithName: {
          name: 'myDataSource',
          type: 'NONE',
        },
        myOtherDataSource: {
          type: 'NONE',
        },
      },
    });
    expect(config.dataSources).toMatchSnapshot();
  });

  it('should merge dataSource arrays', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      dataSources: [
        {
          dataSourceWithName: {
            name: 'myDataSource',
            type: 'NONE',
          },
          myOtherDataSource: {
            type: 'NONE',
          },
        },
        {
          otherSource: {
            name: 'otherNamedDs',
            type: 'NONE',
          },
          anotherNamedSource: {
            type: 'NONE',
          },
        },
      ],
    });
    expect(config.dataSources).toMatchSnapshot();
  });
});

describe('Resolvers', () => {
  it('should resolve resolver type and fields', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      resolvers: {
        'Query.getUser': {
          dataSource: 'users',
        },
        getUsersResolver: {
          type: 'Query',
          field: 'getUsers',
          dataSource: 'users',
        },
      },
    });
    expect(config.resolvers).toMatchSnapshot();
  });

  it('should merge resolvers arrays', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      resolvers: [
        {
          'Query.getUser': {
            dataSource: 'users',
          },
          getUsersResolver: {
            type: 'Query',
            field: 'getUsers',
            dataSource: 'users',
          },
          'Query.pipeline': {
            kind: 'PIPELINE',
            functions: ['function1', 'function2'],
          },
        },
        {
          'Query.getPosts': {
            dataSource: 'posts',
          },
          getPostsResolver: {
            type: 'Query',
            field: 'getPosts',
            dataSource: 'posts',
          },
          pipelineResolver2: {
            kind: 'PIPELINE',
            functions: ['function1', 'function2'],
            type: 'Query',
            field: 'getUsers',
          },
        },
      ] as Record<string, ResolverConfigInput>[],
    });
    expect(config.resolvers).toMatchSnapshot();
  });
});

describe('Pipeline Functions', () => {
  it('should resolve functions names', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      pipelineFunctions: {
        function1: {
          dataSource: 'users',
        },
        function2: {
          name: 'myFunction2',
          dataSource: 'users',
        },
      },
    });
    expect(config.pipelineFunctions).toMatchSnapshot();
  });

  it('should merge function arrays', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,

      pipelineFunctions: [
        {
          function1: {
            dataSource: 'users',
          },
          function2: {
            name: 'myFunction2',
            dataSource: 'users',
          },
        },
        {
          function3: {
            dataSource: 'users',
          },
          function4: {
            name: 'myFunction4',
            dataSource: 'users',
          },
        },
      ],
    });
    expect(config.pipelineFunctions).toMatchSnapshot();
  });
});
