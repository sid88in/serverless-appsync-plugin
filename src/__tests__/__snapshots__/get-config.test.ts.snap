// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema as absolute path 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  meInfo(consumer_key: String, consumer_secret: String): User!
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;

exports[`Schema as array 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  meInfo(consumer_key: String, consumer_secret: String): User!
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;

exports[`Schema as glob pattern 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  meInfo(consumer_key: String, consumer_secret: String): User!
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;

exports[`Schema as string 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  meInfo(consumer_key: String, consumer_secret: String): User!
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;

exports[`datasources as array 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [
      Object {
        "name": "users",
        "type": "AMAZON_DYNAMODB",
      },
      Object {
        "name": "tweets",
        "type": "AMAZON_DYNAMODB",
      },
    ],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  meInfo(consumer_key: String, consumer_secret: String): User!
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;

exports[`datasources as array form different files (array of arrays or objects) 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [
      Object {
        "name": "users",
        "type": "AMAZON_DYNAMODB",
      },
      Object {
        "name": "tweets",
        "type": "AMAZON_DYNAMODB",
      },
      Object {
        "name": "foo",
        "type": "AMAZON_DYNAMODB",
      },
      Object {
        "name": "bar",
        "type": "AMAZON_DYNAMODB",
      },
    ],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  meInfo(consumer_key: String, consumer_secret: String): User!
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;

exports[`returns valid config 1`] = `
Array [
  Object {
    "additionalAuthenticationProviders": Array [],
    "authentication": Object {
      "type": "AWS_IAM",
    },
    "dataSources": Array [
      Object {
        "name": "users",
        "type": "AMAZON_DYNAMODB",
      },
      Object {
        "name": "tweets",
        "type": "AMAZON_DYNAMODB",
      },
    ],
    "defaultMappingTemplates": Object {},
    "isSingleConfig": true,
    "mappingTemplatesLocation": Object {
      "pipelineFunctions": "mapping-templates",
      "resolvers": "mapping-templates",
    },
    "name": "api",
    "pipelineFunctions": Array [],
    "region": "us-east-1",
    "resolvers": Array [],
    "schema": "type Mutation  {
  createTweet(tweet: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String, created_at: String!): Tweet!
  deleteTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  reTweet(tweet_id: String!, consumer_key: String, consumer_secret: String, access_token_key: String, access_token_secret: String): Tweet!
  updateTweet(tweet_id: String!, tweet: String!): Tweet!
  updateUserInfo(location: String!, description: String!, name: String!, followers_count: Int!, friends_count: Int!, favourites_count: Int!, followers: [String!]!): User!
}

type Query  {
  meInfo(consumer_key: String, consumer_secret: String): User!
  getUserInfo(handle: String!, consumer_key: String, consumer_secret: String): User!
  searchAllTweetsByKeyword(keyword: String!): TweetConnection
}

type Subscription  {
  addTweet: Tweet @aws_subscribe(mutations: [\\"createTweet\\"])
}

type Tweet  {
  tweet_id: String!
  tweet: String!
  retweeted: Boolean
  retweet_count: Int
  favorited: Boolean
  created_at: String!
}

type TweetConnection  {
  items: [Tweet!]!
  nextToken: String
}

type User  {
  name: String!
  handle: String!
  location: String!
  description: String!
  followers_count: Int!
  friends_count: Int!
  favourites_count: Int!
  followers: [String!]!
  topTweet: Tweet
  tweets(limit: Int!, nextToken: String): TweetConnection
  searchTweetsByKeyword(keyword: String!): TweetConnection
}",
    "substitutions": Object {},
    "xrayEnabled": false,
  },
]
`;
